File: analyzer-direct/src/main/kotlin/io/github/tassiLuca/analyzerkt/lib/Analyzer.kt
package io.github.tassiLuca.analyzerkt.lib

import kotlinx.coroutines.async
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch

/** A generic analyzer of organization/group/workspace repositories. */
interface Analyzer {

    /**
     * Performs a **suspending** analysis of the [organizationName]'s repositories, providing the results
     * incrementally to the [updateResults] function, along with the indication of the completion.
     */
    suspend fun analyze(
        organizationName: String,
        updateResults: suspend (RepositoryReport) -> Unit = { _ -> },
    ): Result<Set<RepositoryReport>>

    companion object {
        /** Creates a new GitHub organization [Analyzer]. */
        fun ofGitHub(): Analyzer = GitHubAnalyzer()
    }
}

private class GitHubAnalyzer : Analyzer {

    private val gitHubProvider = GitHubRepositoryProvider(GitHubService.create())

    override suspend fun analyze(
        organizationName: String,
        updateResults: suspend (RepositoryReport) -> Unit,
    ): Result<Set<RepositoryReport>> = coroutineScope {
        runCatching {
            val channel = Channel<RepositoryReport>()
            val repositories = gitHubProvider.repositoriesOf(organizationName).getOrThrow()
            repositories.forEach {
                launch {
                    val contributors = async { gitHubProvider.contributorsOf(organizationName, it.name).getOrThrow() }
                    val release = async { gitHubProvider.lastReleaseOf(organizationName, it.name).getOrThrow() }
                    channel.send(RepositoryReport(it.name, it.issues, it.stars, contributors.await(), release.await()))
                }
            }
            var allReports = emptySet<RepositoryReport>()
            repeat(repositories.size) {
                val report = channel.receive()
                allReports = allReports + report
                updateResults(report)
            }
            channel.close()
            allReports
        }
    }
}

File: analyzer-monadic/src/main/scala/io/github/tassiLuca/analyzer/lib/Analyzer.scala
package io.github.tassiLuca.analyzer.lib

import cats.data.EitherT
import io.github.tassiLuca.analyzer.commons.lib
import io.github.tassiLuca.analyzer.commons.lib.{Repository, RepositoryReport}
import monix.eval.Task

/** A generic analyzer of organization/group/workspace repositories. */
trait Analyzer:

  /** @return a [[EitherT]] encapsulating a [[Task]] that performs the analysis of the
    * [[organizationName]]'s repositories, providing the results incrementally to the
    * [[updateResults]] function.
    */
  def analyze(organizationName: String)(
      updateResult: RepositoryReport => Unit,
  ): EitherT[Task, String, Seq[RepositoryReport]]

object Analyzer:
  def apply(repositoryService: RepositoryService): Analyzer = AnalyzerImpl(repositoryService)

  private class AnalyzerImpl(repositoryService: RepositoryService) extends Analyzer:

    import cats.implicits.catsSyntaxParallelTraverse1

    override def analyze(organizationName: String)(
        updateResult: RepositoryReport => Unit,
    ): EitherT[Task, String, Seq[RepositoryReport]] =
      for
        repositories <- repositoryService.repositoriesOf(organizationName)
        reports <- repositories.parTraverse(r => EitherT.right(r.performAnalysis(updateResult)))
      yield reports

    extension (r: Repository)
      private def performAnalysis(updateResult: RepositoryReport => Unit): Task[RepositoryReport] =
        val contributorsTask = repositoryService.contributorsOf(r.organization, r.name).value
        val releaseTask = repositoryService.lastReleaseOf(r.organization, r.name).value
        for
          result <- Task.parZip2(contributorsTask, releaseTask)
          report = RepositoryReport(r.name, r.issues, r.stars, result._1.getOrElse(Seq.empty), result._2.toOption)
          _ <- Task(updateResult(report))
        yield report
